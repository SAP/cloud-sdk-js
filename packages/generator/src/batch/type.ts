import { StructureKind, TypeAliasDeclarationStructure } from 'ts-morph';
import { VdmServiceMetadata } from '../vdm-types';
import { operationReturnType } from '../operations';

/**
 * @internal
 */
export function readRequestType(
  service: VdmServiceMetadata
): TypeAliasDeclarationStructure {
  return {
    kind: StructureKind.TypeAlias,
    name: `Read${service.className}RequestBuilder<DeSerializersT extends DeSerializers>`,
    isExported: true,
    type: getReadRequestType(service)
  };
}
/**
 * @internal
 */
export function writeRequestType(
  service: VdmServiceMetadata
): TypeAliasDeclarationStructure {
  return {
    kind: StructureKind.TypeAlias,
    name: `Write${service.className}RequestBuilder<DeSerializersT extends DeSerializers>`,
    isExported: true,
    type: getWriteRequestType(service)
  };
}

function getWriteRequestType(service: VdmServiceMetadata): string {
  const createUpdateDeleteBuilderTypes = service.entities.map(
    e =>
      `CreateRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT> | UpdateRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT> | DeleteRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );
  const functionImportsReturnTypes = service.functionImports
    .filter(fn => fn.httpMethod.toLowerCase() !== 'get')
    .map(fn => operationReturnType(fn));
  const actionImportsReturnTypes =
    service.actionImports?.map(fn => operationReturnType(fn)) ?? [];
  return [
    ...createUpdateDeleteBuilderTypes,
    ...functionImportsReturnTypes,
    ...actionImportsReturnTypes
  ].join(' | ');
}

function getReadRequestType(service: VdmServiceMetadata): string {
  const getAllBuilderTypes = service.entities.map(
    e => `GetAllRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );

  const getByKeyBuilderTypes = service.entities.map(
    e =>
      `GetByKeyRequestBuilder<${e.className}<DeSerializersT>, DeSerializersT>`
  );
  const functionImportsReturnTypes = service.functionImports
    .filter(fn => fn.httpMethod.toLowerCase() === 'get')
    .map(fn => operationReturnType(fn));

  return [
    ...getAllBuilderTypes,
    ...getByKeyBuilderTypes,
    ...functionImportsReturnTypes
  ].join(' | ');
}
