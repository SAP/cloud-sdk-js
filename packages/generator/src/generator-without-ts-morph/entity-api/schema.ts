import { documentationBlock } from '@sap-cloud-sdk/util';
import { VdmEntity, VdmProperty } from '../../vdm-types';
import { addLeadingNewline, getStaticPropertyDescription } from '../../typedoc';
import { createPropertyFieldInitializerForEntity } from '../../entity';

/**
 * @internal
 */
export function getSchema(entity: VdmEntity): string {
  return `{ 
    ${[
      ...properties(entity),
      '...this.navigationPropertyFields',
      // ...navigationProperties(entity, service),
      allFieldsSelector(entity)
    ].join(',\n')} 
  }`;
}

function properties(entity: VdmEntity): string[] {
  return entity.properties.map(prop => property(prop));
}

function property(prop: VdmProperty): string {
  return [
    documentationBlock`${getStaticPropertyDescription(prop)}`,
    `${prop.staticPropertyName}: ${createPropertyFieldInitializerForEntity(
      prop,
      'fieldBuilder'
    )}`
  ].join('\n');
}

function allFieldsSelector(entity: VdmEntity): string {
  return [
    documentationBlock`${addLeadingNewline('All fields selector.')}`,
    `ALL_FIELDS: new AllFields('*', ${entity.className})`
  ].join('\n');
}
