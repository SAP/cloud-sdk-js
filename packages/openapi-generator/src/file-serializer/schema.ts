import { codeBlock, documentationBlock, unixEOL } from '@sap-cloud-sdk/util';
import {
  OpenApiSchema,
  OpenApiObjectSchema,
  OpenApiObjectSchemaProperty
} from '../openapi-types';
import { getType } from '../parser/type-mapping';
import {
  isReferenceObject,
  isArraySchema,
  isObjectSchema,
  isEnumSchema,
  isOneOfSchema,
  isAllOfSchema,
  isAnyOfSchema,
  isNotSchema,
  getSchemaPropertiesDocumentation
} from '../schema-util';

/**
 * Serialize a schema.
 * @param schema - Parsed schema to be serialized.
 * @returns The serialized schema
 * @internal
 */
export function serializeSchema(schema: OpenApiSchema): string {
  if (isReferenceObject(schema)) {
    return schema.schemaName;
  }
  if (isArraySchema(schema)) {
    const type = serializeSchema(schema.items);
    return schema.uniqueItems ? `Set<${type}>` : `${type}[]`;
  }
  if (isObjectSchema(schema)) {
    return serializeObjectSchema(schema);
  }

  if (isEnumSchema(schema)) {
    return schema.enum.join(' | ');
  }

  if (isOneOfSchema(schema)) {
    return schema.oneOf.map(type => serializeSchema(type)).join(' | ');
  }

  if (isAllOfSchema(schema)) {
    return schema.allOf.map(type => serializeSchema(type)).join(' & ');
  }

  if (isAnyOfSchema(schema)) {
    return schema.anyOf.map(type => serializeSchema(type)).join(' | ');
  }

  if (isNotSchema(schema)) {
    return codeBlock`any`;
  }

  return getType(schema.type);
}

function serializeObjectSchema(schema: OpenApiObjectSchema): string {
  if (
    !schema.properties.length &&
    typeof schema.additionalProperties !== 'object'
  ) {
    return 'Record<string, any>';
  }

  const types: string[] = [];
  if (schema.properties.length) {
    types.push(serializeObjectSchemaForProperties(schema.properties));
  }

  if (schema.additionalProperties) {
    types.push(
      `Record<string, ${serializeSchema(schema.additionalProperties)}>`
    );
  }

  return types.join(' | ');
}

function serializeObjectSchemaForProperties(
  properties: OpenApiObjectSchemaProperty[]
): string {
  return codeBlock`{
      ${properties
        .map(property => serializePropertyWithDocumentation(property))
        .join(unixEOL)}
    }`;
}

function serializeProperty(name: string, required: boolean, type: string) {
  return `'${name}'${required ? '' : '?'}: ${type};`;
}

function serializePropertyWithDocumentation(
  property: OpenApiObjectSchemaProperty
) {
  const documentation = schemaPropertyDocumentation(property);
  const serialized = serializeProperty(
    property.name,
    property.required,
    serializeSchema(property.schema)
  );
  if (documentation) {
    return [documentation, serialized].join(unixEOL);
  }
  return serialized;
}

/**
 * @internal
 */
export function schemaPropertyDocumentation(
  schema: OpenApiObjectSchemaProperty
): string {
  const signature: string[] = [];
  if (schema.description) {
    signature.push(schema.description);
  }

  signature.push(...getSchemaPropertiesDocumentation(schema.schemaProperties));

  return documentationBlock`${signature.join(unixEOL)}`;
}
