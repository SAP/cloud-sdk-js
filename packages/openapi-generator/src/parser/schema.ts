import { createLogger } from '@sap-cloud-sdk/util';
import { OpenAPIV3 } from 'openapi-types';
import { isReferenceObject } from '../schema-util';
import {
  OpenApiArraySchema,
  OpenApiEnumSchema,
  OpenApiObjectSchema,
  OpenApiObjectSchemaProperty,
  OpenApiReferenceSchema,
  OpenApiSchema,
  OpenApiSchemaProperties
} from '../openapi-types';
import { getType } from './type-mapping';
import { OpenApiDocumentRefs } from './refs';
import { ParserOptions } from './options';

const logger = createLogger('openapi-generator');

/**
 * Parse the original schema or reference object to a serializable schema.
 * @param schema - Originally provided schema or reference object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema.
 * @internal
 */
export function parseSchema(
  schema: OpenAPIV3.SchemaObject | OpenAPIV3.ReferenceObject | undefined,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiSchema {
  if (!schema) {
    logger.verbose("No schema provided, continuing with 'any'.");
    return { type: 'any' };
  }

  if (isReferenceObject(schema)) {
    return parseReferenceSchema(schema, refs);
  }

  if (schema.type === 'array') {
    return parseArraySchema(schema, refs, options);
  }

  if (
    schema.type === 'object' ||
    schema.properties ||
    'additionalProperties' in schema
  ) {
    return parseObjectSchema(schema, refs, options);
  }

  if (schema.enum?.length) {
    return parseEnumSchema(schema, options);
  }

  if (schema.oneOf?.length) {
    return parseXOfSchema(schema, refs, 'oneOf', options);
  }

  if (schema.allOf?.length) {
    return parseXOfSchema(schema, refs, 'allOf', options);
  }

  if (schema.anyOf?.length) {
    return parseXOfSchema(schema, refs, 'anyOf', options);
  }

  if (schema.not) {
    return {
      not: parseSchema(schema.not, refs, options)
    };
  }

  return {
    type: getType(schema.type)
  };
}

function parseReferenceSchema(
  schema: OpenAPIV3.ReferenceObject,
  refs: OpenApiDocumentRefs
): OpenApiReferenceSchema {
  return {
    ...schema,
    ...refs.getSchemaNaming(schema)
  };
}

/**
 * Parse a schema to an array schema.
 * @param schema - Original schema representing an array.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed array schema.
 */
function parseArraySchema(
  schema: OpenAPIV3.ArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiArraySchema {
  return {
    uniqueItems: schema.uniqueItems,
    items: parseSchema(schema.items, refs, options)
  };
}

/**
 * Parse a schema to an object schema.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The recursively parsed object schema.
 */
function parseObjectSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiObjectSchema {
  const properties = parseObjectSchemaProperties(schema, refs, options);

  if (schema.additionalProperties === false) {
    if (!properties.length) {
      throw new Error(
        'Could not parse object schema without neither properties nor additional properties.'
      );
    }

    return { properties };
  }

  const additionalProperties =
    typeof schema.additionalProperties === 'object' &&
    Object.keys(schema.additionalProperties).length
      ? parseSchema(schema.additionalProperties, refs, options)
      : { type: 'any' };

  return {
    properties,
    additionalProperties
  };
}

/**
 * Parse properties of an object as property schemas.
 * @param schema - Original schema representing an object.
 * @param refs - Object representing cross references throughout the document.
 * @param options - Options that were set for service generation.
 * @returns The list of parsed property schemas.
 */
function parseObjectSchemaProperties(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  options: ParserOptions
): OpenApiObjectSchemaProperty[] {
  return Object.entries(schema.properties || {}).reduce(
    (props, [propName, propSchema]) => [
      ...props,
      {
        schema: parseSchema(propSchema, refs, options),
        description: isReferenceObject(propSchema)
          ? undefined
          : propSchema.description,
        name: propName,
        required: schema.required?.includes(propName) || false,
        schemaProperties: { ...parseSchemaProperties(propSchema) }
      }
    ],
    []
  );
}

/**
 * Parse an enum schema.
 * @param schema - Original schema representing an enum.
 * @param options - Options that were set for service generation.
 * @returns The parsed enum schema.
 */
function parseEnumSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  options: ParserOptions
): OpenApiEnumSchema {
  const type = schema.type ? getType(schema.type) : 'string';
  return {
    type,
    enum: (schema.enum || []).map(entry => {
      if (type === 'string' && entry !== null) {
        return getEnumStringValue(String(entry));
      }
      if (entry === null && !schema.nullable) {
        if (options.strictNaming) {
          throw new Error(
            'null was used as a parameter in an enum, although the schema was not declared as nullable'
          );
        } else {
          logger.warn(
            'null was used as a parameter in an enum, although the schema was not declared as nullable'
          );
        }
      }
      return entry;
    })
  };
}

function getEnumStringValue(input: string): string {
  return `'${input.replace(/'/g, "\\'")}'`;
}

/**
 * Parse a 'oneOf', 'allOf' or 'anyOf' schema.
 * @param schema - Original schema to parse.
 * @param refs - Object representing cross references throughout the document.
 * @param xOf - Key to identify which schema to parse.
 * @param options - Options that were set for service generation.
 * @returns The parsed schema based on the given key.
 */
function parseXOfSchema(
  schema: OpenAPIV3.NonArraySchemaObject,
  refs: OpenApiDocumentRefs,
  xOf: 'oneOf' | 'allOf' | 'anyOf',
  options: ParserOptions
): any {
  return {
    [xOf]: (schema[xOf] || []).map(entry => parseSchema(entry, refs, options))
  };
}

/**
 * Parse schema properties e.g. 'maxLength', 'minimum', etc.
 * @param schema - Original schema representing a ref or schema object.
 * @returns The parsed schema properties object.
 * @internal
 */
export function parseSchemaProperties(
  schema: OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject
): OpenApiSchemaProperties {
  if (isReferenceObject(schema)) {
    return {};
  }
  const schemaPropertyNames = [
    'deprecated',
    'example',
    'format',
    'default',
    'multipleOf',
    'maximum',
    'minimum',
    'maxLength',
    'minLength',
    'minItems',
    'maxItems',
    'pattern'
  ];
  return schemaPropertyNames.reduce((properties, propertyName) => {
    if (schema[propertyName]) {
      return { ...properties, [propertyName]: schema[propertyName] };
    }
    return properties;
  }, {});
}
