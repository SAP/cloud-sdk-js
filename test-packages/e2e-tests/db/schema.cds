namespace my.namespace;

entity TestEntity {
  key KeyPropertyInt: Integer;
  key KeyPropertyString: String;
  StringProperty: String(111);
  GuidProperty: UUID;
  BooleanProperty: Boolean;
  Int64Property: Integer64;
  DoubleProperty: Double;
  DecimalProperty: Decimal(9,2);
  DateProperty: Date;
  TimeOfDayProperty: Time;
  DataTimeOffsetDataTimeProperty: DateTime;
  DataTimeOffsetTimestampProperty: Timestamp;
}
