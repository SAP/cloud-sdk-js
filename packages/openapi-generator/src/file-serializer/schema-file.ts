import { codeBlock, documentationBlock, unixEOL } from '@sap-cloud-sdk/util';
import {
  serializeImports,
  Import
} from '@sap-cloud-sdk/generator-common/internal';
import { OpenApiPersistedSchema } from '../openapi-types';
import { collectRefs, getSchemaPropertiesDocumentation } from '../schema-util';
import { serializeSchema } from './schema';

/**
 * Serialize a schema representation to a string representing the according schema file contents.
 * @param namedSchema - A named schema.
 * @returns The serialized schema file contents.
 * @internal
 */
export function schemaFile(namedSchema: OpenApiPersistedSchema): string {
  const imports = serializeImports(getImports(namedSchema));

  return codeBlock`    
    ${imports}
    ${schemaDocumentation(namedSchema)}
    export type ${namedSchema.schemaName} = ${serializeSchema(
    namedSchema.schema
  )};
  `;
}

function getImports(namedSchema: OpenApiPersistedSchema): Import[] {
  return collectRefs(namedSchema.schema)
    .filter(ref => ref.schemaName !== namedSchema.schemaName)
    .map(ref => ({
      names: [ref.schemaName],
      typeOnly: true,
      moduleIdentifier: `./${ref.fileName}`
    }));
}

/**
 * @internal
 */
export function schemaDocumentation(schema: OpenApiPersistedSchema): string {
  const lines = [
    schema.description ||
      `Representation of the '${schema.schemaName}' schema.`,
    ...getSchemaPropertiesDocumentation(schema.schemaProperties)
  ];
  return documentationBlock`${lines.join(unixEOL)}`;
}
