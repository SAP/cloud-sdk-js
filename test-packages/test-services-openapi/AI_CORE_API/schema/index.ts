/*
 * Copyright (c) 2024 SAP SE or an SAP affiliate company. All rights reserved.
 *
 * This is a generated file powered by the SAP Cloud SDK for JavaScript.
 */
export * from './url';
export * from './api-error-with-id';
export * from './artifact-list';
export * from './resource-list';
export * from './artifact-array';
export * from './artifact';
export * from './artifact-base-data';
export * from './labeled';
export * from './label-list';
export * from './label';
export * from './label-key';
export * from './label-value';
export * from './artifact-name';
export * from './generic-name';
export * from './generic-natural-language-term';
export * from './artifact-url';
export * from './artifact-description';
export * from './artifact-detail-data';
export * from './artifact-id';
export * from './id';
export * from './scenario-id';
export * from './configuration-id';
export * from './execution-id';
export * from './creation-data';
export * from './modification-data';
export * from './expanded-scenario';
export * from './scenario';
export * from './scenario-base-data';
export * from './scenario-label-list';
export * from './scenario-label';
export * from './scenario-label-key';
export * from './scenario-detail-data';
export * from './artifact-creation-response';
export * from './creation-response';
export * from './creation-response-message';
export * from './message';
export * from './artifact-creation-response-message';
export * from './artifact-post-data';
export * from './api-error';
export * from './configuration-list';
export * from './configuration';
export * from './configuration-base-data';
export * from './configuration-name';
export * from './executable-id';
export * from './parameter-argument-binding-list';
export * from './parameter-argument-binding';
export * from './artifact-argument-binding-list';
export * from './artifact-argument-binding';
export * from './configuration-detail-data';
export * from './configuration-creation-response';
export * from './configuration-creation-response-message';
export * from './deployment-creation-response';
export * from './deployment-url';
export * from './execution-status';
export * from './status';
export * from './deployment-creation-response-message';
export * from './deployment-time-to-live';
export * from './execution-creation-response';
export * from './execution-creation-response-message';
export * from './deployment-list';
export * from './deployment';
export * from './deployment-detail-data';
export * from './deployment-id';
export * from './deployment-status';
export * from './deployment-status-message';
export * from './status-transition-time';
export * from './deployment-creation-request';
export * from './enactment-creation-request';
export * from './deployment-bulk-modification-response';
export * from './deployment-modification-response-list';
export * from './deployment-bulk-modification-request';
export * from './deployment-modification-request-list';
export * from './deployment-modification-request-with-identifier';
export * from './deployment-response-with-details';
export * from './deployment-detail-properties';
export * from './deployment-status-details';
export * from './deployment-details';
export * from './scaling-details';
export * from './backend-details';
export * from './resources-details';
export * from './deployment-deletion-response';
export * from './deployment-deletion-response-message';
export * from './deployment-modification-response';
export * from './deployment-modification-response-message';
export * from './deployment-modification-request';
export * from './deployment-target-status';
export * from './execution-list';
export * from './execution';
export * from './execution-data';
export * from './execution-status-message';
export * from './execution-schedule-id';
export * from './execution-bulk-modification-response';
export * from './execution-modification-response-list';
export * from './execution-bulk-modification-request';
export * from './execution-modification-request-list';
export * from './execution-modification-request-with-identifier';
export * from './execution-response-with-details';
export * from './execution-status-details';
export * from './execution-deletion-response';
export * from './execution-deletion-response-message';
export * from './execution-modification-response';
export * from './execution-modification-response-message';
export * from './execution-modification-request';
export * from './execution-schedule-list';
export * from './execution-schedule';
export * from './execution-schedule-data';
export * from './execution-schedule-creation-data';
export * from './cron';
export * from './schedule-start-end';
export * from './execution-schedule-status';
export * from './execution-schedule-creation-response';
export * from './execution-schedule-creation-response-message';
export * from './execution-schedule-deletion-response';
export * from './execution-schedule-deletion-response-message';
export * from './execution-schedule-modification-response';
export * from './execution-schedule-modification-response-message';
export * from './execution-schedule-modification-request';
export * from './schedule';
export * from './scenario-list';
export * from './executable-list';
export * from './executable';
export * from './executable-base-data';
export * from './executable-detail-data';
export * from './version-id';
export * from './executable-parameter-list';
export * from './executable-parameter';
export * from './executable-artifact-list';
export * from './executable-artifact';
export * from './version-list';
export * from './version';
export * from './version-base-data';
export * from './version-description';
export * from './version-detail-data';
export * from './capabilities';
export * from './version-2';
export * from './ai-api';
export * from './spec';
export * from './extensions';
export * from './get-metric-resource-list';
export * from './resource-count';
export * from './get-metric-resource';
export * from './metric-resource';
export * from './execution-id-2';
export * from './mlapi-execution-id';
export * from './metric-list';
export * from './metric';
export * from './metric-name';
export * from './metric-value';
export * from './timestamp';
export * from './label-list-2';
export * from './label-2';
export * from './label-name';
export * from './tag-list';
export * from './tag';
export * from './generic-name-2';
export * from './custom-info-object-list';
export * from './custom-info-object';
export * from './custom-info-object-data';
export * from './get-metric-list';
export * from './get-metric';
export * from './delete-metrics-response';
export * from './string-array';
export * from './metric-selector-permissible-values';
export * from './api-error-2';
export * from './details-error-response';
export * from './result-set';
export * from './array-of-column-names';
export * from './column-name';
export * from './result-row-list';
export * from './result-row';
export * from './result-row-item';
export * from './aggregation-attribute';
export * from './count-aggregate';
export * from './error';
export * from './file-creation-response';
export * from './object-store-secret-status-response';
export * from './object-store-secret-status';
export * from './object-store-secret-creation-response';
export * from './object-store-secret-creation-response-message';
export * from './error-response';
export * from './error-2';
export * from './object-store-secret-with-sensitive-data-request-for-post-call';
export * from './object-store-secret-deletion-response';
export * from './creation-response-2';
export * from './object-store-secret-deletion-response-message';
export * from './object-store-secret-modification-response';
export * from './object-store-secret-modification-response-message';
export * from './object-store-secret-with-sensitive-data-request';
export * from './list-generic-secrets-response';
export * from './generic-secret-details';
export * from './generic-secret-data-response';
export * from './name';
export * from './generic-secret-post-body';
export * from './generic-secret-data';
export * from './generic-secret-patch-body';
export * from './resource-group-list';
export * from './resource-group';
export * from './resource-group-base';
export * from './resource-group-labels';
export * from './resource-group-label';
export * from './resource-groups-post-request';
export * from './resource-group-deletion-response';
export * from './resource-group-deletion-response-message';
export * from './resource-group-patch-request';
export * from './resource-get-response';
export * from './resource-get-resource-plans';
export * from './resource-patch-response';
export * from './resource-patch-body';
export * from './resource-patch-resource-plans';
export * from './resource-patch-nodes';
export * from './log-common-response';
export * from './log-common-data';
export * from './log-common-result';
export * from './log-common-result-item';
export * from './timestamp-1';
export * from './message-1';
export * from './error-response-2';
export * from './argo-cd-repository-data-response';
export * from './argo-cd-repository-details';
export * from './argo-cd-repository-creation-response';
export * from './argo-cd-repository-creation-response-message';
export * from './argo-cd-repository-data';
export * from './argo-cd-repository-deletion-response';
export * from './argo-cd-repository-deletion-response-message';
export * from './argo-cd-repository-modification-response';
export * from './argo-cd-repository-modification-response-message';
export * from './argo-cd-repository-credentials';
export * from './all-argo-cd-application-data';
export * from './argo-cd-application-data';
export * from './argo-cd-application-base-data';
export * from './argo-cd-application-creation-response';
export * from './argo-cd-application-creation-response-message';
export * from './argo-cd-application-data-repo-name';
export * from './argo-cd-application-status';
export * from './argo-cd-application-deletion-response';
export * from './argo-cd-application-deletion-response-message';
export * from './argo-cd-application-modification-response';
export * from './argo-cd-application-modification-response-message';
export * from './argo-cd-application-refresh-response';
export * from './argo-cd-application-refresh-response-message';
export * from './docker-registry-secret-status';
export * from './docker-registry-secret-deletion-response';
export * from './docker-registry-secret-deletion-response-message';
export * from './docker-registry-secret-modification-response';
export * from './docker-registry-secret-modification-response-message';
export * from './docker-registry-secret-with-sensitive-data-request';
export * from './docker-registry-secret-status-response';
export * from './docker-registry-secret-creation-response';
export * from './docker-registry-secret-creation-response-message';
export * from './docker-registry-name-component';
export * from './service-list';
export * from './service';
export * from './service-base';
export * from './extended-service';
export * from './service-details';
export * from './service-broker-secret';
export * from './service-capabilities';
export * from './service-service-catalog';
export * from './service-service-catalog-item';
export * from './service-service-catalog-item-extend-catalog';
export * from './service-service-plan-item';
export * from './service-service-plan-item-metadata';
export * from './service-service-catalog-item-extend-credentials';
export * from './name-1';
export * from './inline-response-400';
export * from './inline-response-4001';
export * from './body';
export * from './body-1';
export * from './ai-api-capabilities-logs';
export * from './ai-api-capabilities-bulk-updates';
export * from './ai-api-capabilities';
export * from './ai-api-limits-executions';
export * from './ai-api-limits-deployments';
export * from './ai-api-limits-time-to-live-deployments';
export * from './ai-api-limits';
export * from './error-details';
export * from './object-store-secret-status-metadata';
export * from './argo-cd-application-status-source';
export * from './argo-cd-application-status-sync-resources-status';
export * from './service-capabilities-logs';
export * from './service-capabilities-basic';
export * from './service-service-catalog-item-extend-credentials-shared-service-urls';
export * from './service-service-catalog-item-extend-credentials-shared';
