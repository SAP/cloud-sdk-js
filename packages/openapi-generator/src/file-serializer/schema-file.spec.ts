import {
  OpenApiObjectSchemaProperty,
  OpenApiPersistedSchema
} from '../openapi-types';
import { schemaDocumentation, schemaFile } from './schema-file';
import { schemaPropertyDocumentation } from './schema';
describe('schemaFile', () => {
  it('serializes schema file for schema', () => {
    expect(
      schemaFile({
        schemaName: 'MySchema',
        fileName: 'my-schema',
        schema: {
          properties: [
            {
              name: 'string-property',
              required: true,
              schema: {
                type: 'string'
              },
              schemaProperties: {
                maxLength: 10
              }
            }
          ]
        },
        schemaProperties: {
          deprecated: true
        }
      })
    ).toMatchInlineSnapshot(`
      "    
          /**
           * Representation of the 'MySchema' schema.
           * @deprecated
           */
          export type MySchema = {
            /**
             * Max Length: 10.
             */
            'string-property': string;
          };"
    `);
  });

  it('serializes schema file for schema including references', () => {
    expect(
      schemaFile({
        schemaName: 'MySchema',
        fileName: 'my-schema',
        schemaProperties: {},
        schema: {
          properties: [
            {
              name: 'otherSchema1',
              required: true,
              schema: {
                $ref: '#/components/schema/OtherSchema1',
                schemaName: 'OtherSchema1',
                fileName: 'other-schema-1'
              },
              schemaProperties: {}
            },
            {
              name: 'otherSchema2',
              description: 'Description other Schema 2',
              required: true,
              schema: {
                $ref: '#/components/schema/OtherSchema2',
                schemaName: 'OtherSchema2',
                fileName: 'other-schema-2'
              },
              schemaProperties: {}
            }
          ]
        }
      })
    ).toMatchInlineSnapshot(`
      "    import type { OtherSchema1 } from './other-schema-1';
          import type { OtherSchema2 } from './other-schema-2';
          /**
           * Representation of the 'MySchema' schema.
           */
          export type MySchema = {
            'otherSchema1': OtherSchema1;
            /**
             * Description other Schema 2
             */
            'otherSchema2': OtherSchema2;
          };"
    `);
  });

  it('serializes schema file for schema including not schema', () => {
    expect(
      schemaFile({
        schemaName: 'MySchema',
        fileName: 'my-schema',
        schema: {
          items: { not: { type: 'integer' } }
        },
        schemaProperties: {}
      })
    ).toMatchInlineSnapshot(`
      "    
          /**
           * Representation of the 'MySchema' schema.
           */
          export type MySchema = any[];"
    `);
  });

  it('serializes schema file without imports for schema including only self reference', () => {
    expect(
      schemaFile({
        schemaName: 'MySchema',
        fileName: 'my-schema',
        schema: {
          properties: [
            {
              name: 'property',
              required: false,
              schema: {
                $ref: '#/components/schema/MySchema',
                schemaName: 'MySchema',
                fileName: 'my-schema'
              },
              schemaProperties: {}
            }
          ]
        },
        schemaProperties: {}
      })
    ).toMatchInlineSnapshot(`
      "    
          /**
           * Representation of the 'MySchema' schema.
           */
          export type MySchema = {
            'property'?: MySchema;
          };"
    `);
  });

  it('serializes simple schema file for schema with description', () => {
    expect(
      schemaFile({
        schemaName: 'MySchema',
        fileName: 'my-schema',
        schema: {
          properties: [
            {
              name: 'string-property',
              description: 'My description',
              required: true,
              schema: {
                type: 'string'
              },
              schemaProperties: {
                minLength: 2
              }
            },
            {
              name: 'string-property-no-description',
              required: true,
              schema: {
                type: 'string'
              },
              schemaProperties: {}
            }
          ]
        },
        schemaProperties: {}
      })
    ).toMatchInlineSnapshot(`
      "    
          /**
           * Representation of the 'MySchema' schema.
           */
          export type MySchema = {
            /**
             * My description
             * Min Length: 2.
             */
            'string-property': string;
            'string-property-no-description': string;
          };"
    `);
  });

  it('creates schema documentation', () => {
    expect(
      schemaDocumentation({ schemaName: 'mySchema' } as OpenApiPersistedSchema)
    ).toMatchInlineSnapshot(`
      "/**
       * Representation of the 'mySchema' schema.
       */"
    `);
  });

  it('uses the schema description documentation if present', () => {
    expect(
      schemaDocumentation({
        schemaName: 'mySchema',
        description: 'My schema description.'
      } as OpenApiPersistedSchema)
    ).toMatch(/My schema description/);
  });

  it('creates a schema property documentation', () => {
    expect(
      schemaPropertyDocumentation({
        description: 'My property Description.'
      } as OpenApiObjectSchemaProperty)
    ).toMatchInlineSnapshot(`
      "/**
       * My property Description.
       */"
    `);
  });
});
