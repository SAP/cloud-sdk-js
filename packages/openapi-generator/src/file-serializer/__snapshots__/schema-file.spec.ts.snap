// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schemaFile creates a schema property documentation 1`] = `
"/**
 * My property Description.
 */"
`;

exports[`schemaFile creates schema documentation 1`] = `
"/**
 * Representation of the 'mySchema' schema.
 */"
`;

exports[`schemaFile serializes schema file for schema 1`] = `
"    
    /**
     * Representation of the 'MySchema' schema.
     * @deprecated
     */
    export type MySchema = {
      /**
       * Max Length: 10.
       */
      'string-property': string;
    };"
`;

exports[`schemaFile serializes schema file for schema including ESM references 1`] = `
"    import type { OtherSchema1 } from './other-schema-1.js';
    import type { OtherSchema2 } from './other-schema-2.js';
    /**
     * Representation of the 'MySchema' schema.
     */
    export type MySchema = {
      'otherSchema1': OtherSchema1;
      /**
       * Description other Schema 2
       */
      'otherSchema2': OtherSchema2;
    };"
`;

exports[`schemaFile serializes schema file for schema including not schema 1`] = `
"    
    /**
     * Representation of the 'MySchema' schema.
     */
    export type MySchema = any[];"
`;

exports[`schemaFile serializes schema file for schema including references 1`] = `
"    import type { OtherSchema1 } from './other-schema-1';
    import type { OtherSchema2 } from './other-schema-2';
    /**
     * Representation of the 'MySchema' schema.
     */
    export type MySchema = {
      'otherSchema1': OtherSchema1;
      /**
       * Description other Schema 2
       */
      'otherSchema2': OtherSchema2;
    };"
`;

exports[`schemaFile serializes schema file without imports for schema including only self reference 1`] = `
"    
    /**
     * Representation of the 'MySchema' schema.
     */
    export type MySchema = {
      'property'?: MySchema;
    };"
`;

exports[`schemaFile serializes simple schema file for schema with description 1`] = `
"    
    /**
     * Representation of the 'MySchema' schema.
     */
    export type MySchema = {
      /**
       * My description
       * Min Length: 2.
       */
      'string-property': string;
      'string-property-no-description': string;
    };"
`;
